pkg/controller/installation/products/codeready/reconciler.go

```diff
@@ -0,0 +1,385 @@
+	"github.com/sirupsen/logrus"
+	"k8s.io/api/core/v1"
+	k8serr "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	pkgclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var (
+	defaultInstallationNamespace = "codeready-workspaces"
+	defaultClientName            = "che-client"
+	defaultCheClusterName        = "integreatly-cluster"
+	defaultSubscriptionName      = "codeready-workspaces"
+)
+
+func NewReconciler(client pkgclient.Client, rc *rest.Config, coreClient *kubernetes.Clientset, configManager config.ConfigReadWriter, instance *v1alpha1.Installation, logger *logrus.Entry) (*Reconciler, error) {
+	config, err := configManager.ReadCodeReady()
+	if err != nil {
+		return nil, err
+	}
+	if config.GetNamespace() == "" {
+		config.SetNamespace(instance.Spec.NamespacePrefix + defaultInstallationNamespace)
+	}
+	config.SetCreateNamespace(instance.Spec.CreateNamespaces)
+	mpm := marketplace.NewManager(client, rc)
+	return &Reconciler{client: client,
+		coreClient:    coreClient,
+		restConfig:    rc,
+		ConfigManager: configManager,
+		Config:        config,
+		mpm:           mpm,
+		logger:        logger,
+	}, nil
+}
+
+type Reconciler struct {
+	client        pkgclient.Client
+	restConfig    *rest.Config
+	coreClient    *kubernetes.Clientset
+	Config        *config.CodeReady
+	ConfigManager config.ConfigReadWriter
+	mpm           marketplace.MarketplaceInterface
+	logger        *logrus.Entry
+}
+
+func (r *Reconciler) Reconcile(phase v1alpha1.StatusPhase) (v1alpha1.StatusPhase, error) {
+	ctx := context.TODO()
+	switch phase {
+	case v1alpha1.PhaseNone:
+		return r.reconcileNamespace(ctx)
+	case v1alpha1.PhaseAwaitingNS:
+		return r.reconcileNamespace(ctx)
+	case v1alpha1.PhaseCreatingSubscription:
+		return r.reconcileSubscription(ctx)
+	case v1alpha1.PhaseCreatingComponents:
+		return r.reconcileCheCluster(ctx)
+	case v1alpha1.PhaseAwaitingOperator:
+		return r.handleAwaitingOperator(ctx)
+	case v1alpha1.PhaseInProgress:
+		return r.handleProgressPhase(ctx)
+	case v1alpha1.PhaseCompleted:
+		return r.handleReconcile(ctx)
+	case v1alpha1.PhaseFailed:
+		return r.handleReconcile(ctx)
+	default:
+		return phase, nil
+	}
+}
+
+func (r *Reconciler) handleReconcile(ctx context.Context) (v1alpha1.StatusPhase, error) {
+	phase, err := r.reconcileNamespace(ctx)
+	if err != nil {
+		return phase, pkgerr.Wrap(err, "could not reconcile namespace")
+	}
+	phase, err = r.reconcileCheCluster(ctx)
+	if err != nil {
+		return phase, pkgerr.Wrap(err, "could not reconcile checluster")
+	}
+	phase, err = r.reconcileSubscription(ctx)
+	if err != nil {
+		return phase, pkgerr.Wrap(err, "could not reconcile subscription")
+	}
+	phase, err = r.reconcileKeycloakClient(ctx)
+	if err != nil {
+		return phase, pkgerr.Wrap(err, "could not reconcile keycloakrealm")
+	}
+	return phase, nil
+}
+
+func (r *Reconciler) reconcileNamespace(ctx context.Context) (v1alpha1.StatusPhase, error) {
+	if !r.Config.GetCreateNamespace() {
```

Comment:
Is a check for if the namespace exists a better option?
It removes the need for a new CRD property.

```go
	ns := &v1.Namespace{
		ObjectMeta: metav1.ObjectMeta{
			Name: r.Config.GetCreateNamespace(),
		},
	}
	err := r.client.Get(context.TODO(), pkgclient.ObjectKey{Name: r.Config.GetCreateNamespace()}, ns)
	if err != nil && !k8serr.IsNotFound(err) {
		// Create namespace
	}
```

Reply:
From what I understood, the point of the flag suggestion was to stop the operator from creating the namespaces due to permissions issues and instead let them be created by a template externally?

With the current approach, the namespaces can even be created after the `Installation` CR, as the operator will just keep checking if they exist, instead of trying to create them/put stuff into them with unknown permissions?

Or did I miss something here @sedroche @maleck13

Reply:
I guess we still need cluster permissions to `Get` the namespaces, so the property avoids that too.
Im happy to go with this then